e1000: Fix hardcoded BDF

finikorg opened the pull request:
Correct e1000 hardcoded BDF, making it possible to use the same driver when Qemu parameters change.

Fixes #51829

jhedberg commented:
Has someone looked into how the tests should be updated to pass with q35?

henrikbrixandersen commented:
> Has someone looked into how the tests should be updated to pass with q35?

As far as I can see, it is just a matter of where the emulated E1000 is placed on the PCIe bus. Changing `PCIE_BDF(0, 3, 0)` to `PCIE_BDF(0, 2, 0)` should make it work.

henrikbrixandersen reviewed (changes requested):
This will result in different PCIe layouts for `qemu_x86_64` and `qemu_x86`. I suggest just updating the E1000 driver to use the new BDF (or even better, make the driver obtain the BDF from dts).

henrikbrixandersen commented:
I am not talking about PCIe IRQ allocation. I simple refer to https://github.com/zephyrproject-rtos/zephyr/blob/7756809c972d5fff0b52d85cfda5f6f20628a2c2/drivers/ethernet/eth_e1000.c#L238 which hardcodes the BDF for the E1000 driver.

finikorg commented:
> I am not talking about PCIe IRQ allocation. I simple refer to
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/7756809c972d5fff0b52d85cfda5f6f20628a2c2/drivers/ethernet/eth_e1000.c#L238
> which hardcodes the BDF for the E1000 driver.

This can be done. I was thinking to rewrite e1000 using PCIE IRQ allocation.

finikorg committed:
drivers: e1000: Change PCIE BDF for Q35 qemu

Since moving to Qemu Q35 machine the drivers with hardcoded BDF stop
working. Correct e1000 BDF.

Fixes #51829

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

henrikbrixandersen commented:
> This can be done. I was thinking to rewrite e1000 using PCIE IRQ allocation.

That would be nice, but would require dynamic PCIe IRQ allocation supported on non-64 bit x86. I skipped dynamic IRQ allocation for this reason in #51750.

finikorg committed:
qemu_x86: dts: Move e1000 entry from soc to pcie

Use standard PCIE dts representation for e1000 PCI Ethernet driver.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
drivers: e1000: Add BUILD_ASSERT() for IRQ allocation

Add BUILD_ASSERT() to indicate that dynamic IRQ allocation is not
supported.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
drivers: e1000: Use lookup() instead of probe()

For PCI Ethernet driver it makes sense to use lookup(id) instead of
probe(bdf, id). Even when using different Qemu parameters we may get
different BDF for e1000 device.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented on drivers/ethernet/eth_e1000.c:
```diff
@@ -230,18 +230,16 @@ static void e1000_isr(const struct device *ddev)
 {
-	const pcie_bdf_t bdf = PCIE_BDF(0, 3, 0);
+	/* PCI ID is decoded into REG_SIZE */
+	pcie_bdf_t bdf = pcie_bdf_lookup(DT_INST_REG_SIZE(0));
```
No, as before

jhedberg commented:
Is the BDF still correct after the CAN support for qemu that was just merged?

finikorg commented:
> Is the BDF still correct after the CAN support for qemu that was just merged?

Should not matter now. [540fecb](https://github.com/zephyrproject-rtos/zephyr/pull/51845/commits/540fecb4f84db7f04a981bf3bd449c3f7eba8801) I think BDF might change depending on Qemu command line options.

finikorg commented:
> > Is the BDF still correct after the CAN support for qemu that was just merged?
> 
> Should not matter now. [540fecb](https://github.com/zephyrproject-rtos/zephyr/pull/51845/commits/540fecb4f84db7f04a981bf3bd449c3f7eba8801) I think BDF might change depending on Qemu command line options.

I will rebase since pcie gets duplicated
